### DISCUSSION TITLE
fmttxt: pixels as formatted text

### DISCUSSION THREAD

**Discussion by snibgo** [ImageMagick/ImageMagick#4515] [Contribute]
I have written a coder for writing and reading pixels as formatted text. It is flexible enough to read the "txt:", "sparse-color:" and "debug:" formats. The main control is a format string, which works in a similar way to "-format" for images. It works just like any other coder, but it can also read and write meta channels.

Documentation, examples and source code are at [Formatted text coder](http://im.snibgo.com/fmttxt.htm).

I welcome any feedback. Please post any questions, comments or criticisms here.

I formally give permission for developers to incorporate it into ImageMagick under the usual "contributor" conditions. It might be wise to wait for any feedback to be resolved.

**Comment by urban-warrior** [comment]
@snibgo, we of course welcome your contribution, in particular, as we need to improve our support of multi-spectral images.  Let us know when you are ready for us to validate fmttxt for inclusion in the ImageMagick distribution.  We will likely make some changes such as adding the traditional coder header with you as the author and it will be distributed under the ImageMagick copyright.  We reserve the right to change the coder name, although we may not.  If you agree to all, post here when your feedback cycle completes.  Thanks.

**Comment by urban-warrior** [comment]
`mgk.c:42:10: fatal error: MagickCore/studio.h: No such file or directory`

Before you build, set the PKG_CONFIG_PATH environment variable if ImageMagick is not in your default system path:

`export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig`

**Comment by snibgo** [comment]
An alternative to "fmttxt:" might be "ftxt:". Feel free to use any name you want.

Feel free to include the header and make any changes you want.

I would appreciate a credit, eg to "snibgo (Alan Gibson)". I don't _require_ a credit.

I am content that the module works as I want. I have had no feedback, so I consider the feedback cycle completed.

The PKG_CONFIG_PATH export doesn't seem to help, but I don't have time right now to investigate further.

**Comment by urban-warrior** [comment]
We merged this coder into the master branch as "ftxt:".  Thanks for contributing this module.

**Comment by fmw42** [comment]
Testing fails for ftxt:

magick rose: -kmeans 5 -unique-colors -depth 8 -define ftxt:format=\H\n ftxt:

Produces 

`HnHnHnHnHn`

 on IM 7.1.0-26.

Am I doing something wrong?   Is this actually implemented in Imagemagick?

But this does produce values:

`magick rose: -kmeans 5 -unique-colors -depth 8 txt:`

```
# ImageMagick pixel enumeration: 5,1,255,srgb
0,0: (65,61,51)  #413D33  srgb(65,61,51)
1,0: (115,125,105)  #737D69  srgb(115,125,105)
2,0: (145,63,50)  #913F32  srgb(145,63,50)
3,0: (223,65,64)  #DF4140  srgb(223,65,64)
4,0: (229,228,230)  #E5E4E6  srgb(229,228,230)

```

**Comment by snibgo** [comment]
That command works fine for me in Windows CMD. But in bash, it needs to be quoted:
```
$ magick.exe rose: -kmeans 5 -unique-colors -depth 8 -define "ftxt:format=\H\n" ftxt:
#413D33
#737D69
#913F32
#DF4140
#E5E4E6
```
Tested with IM v7.1.0-20.

**Comment by observ33r** [comment]
To avoid any problems you can always check if ImageMagic support ftxt format! If not, use "awk" fallback..
```
mapfile -t colors < <(
    [[ -z $(magick -list format | grep -i ftxt) ]] && {
        magick rose: -kmeans 5 -unique-colors -depth 8 \
            txt:- | awk 'NR>1 {print $3}'
    } || {
        magick rose: -kmeans 5 -unique-colors -depth 8 \
            -define ftxt:format="\H\n" ftxt:-
    }
)

declare -p colors
#declare -a colors=([0]="#413D33" [1]="#737D69" [2]="#913F32" [3]="#DF4140" [4]="#E5E4E6")
```
